rift/rib.py

```diff
@@ -19,56 +20,138 @@ def debug(self, msg, *args):
-        packet_common.assert_prefix_address_family(prefix, self.address_family)
-        if prefix in self.destinations:
-            return self.destinations[prefix].get_route(owner)
+    def get_route(self, rte_prefix, owner):
+        packet_common.assert_prefix_address_family(rte_prefix, self.address_family)
+        prefix = str(rte_prefix)
+        if self.destinations.has_key(prefix):
+            return self.destinations.get(prefix).get_route(owner)
         else:
             return None
 
     def put_route(self, rte):
+        """
+        Add a RibRoute object to the Destination object associated to the prefix.
+        :param rte: (RibRoute) the object to add to the Destination associated to the prefix
+        :return:
+        """
+
         packet_common.assert_prefix_address_family(rte.prefix, self.address_family)
         rte.stale = False
-        prefix = rte.prefix
         self.debug("Put %s", rte)
-        if rte.prefix in self.destinations:
-            destination = self.destinations[prefix]
+        # If there is no Destination object for the prefix, create a new Destination object
+        # for the given prefix and insert it in the Trie
+        prefix = packet_common.ip_prefix_str(rte.prefix)
+        if not self.destinations.has_key(prefix):
+            prefix_destination = _Destination(self, rte.prefix)
+            self.destinations.insert(prefix, prefix_destination)
         else:
-            destination = _Destination(prefix)
-            self.destinations[prefix] = destination
-        destination.put_route(rte, self.fib)
-
-    def del_route(self, prefix, owner):
-        # Returns True if the route was present in the table and False if not.
-        packet_common.assert_prefix_address_family(prefix, self.address_family)
-        if prefix in self.destinations:
-            destination = self.destinations[prefix]
-            deleted = destination.del_route(owner, self.fib)
-            if destination.routes == []:
-                del self.destinations[prefix]
+            prefix_destination = self.destinations.get(prefix)
+        # Insert desired route in destination object
+        best_changed = prefix_destination.put_route(rte)
+
+        # Get children prefixes before performing actions on the prefix
+        # (it can be deleted from the Trie)
+        children_prefixes = self.destinations.children(
+            packet_common.ip_prefix_str(prefix_destination.prefix))
+
+        # If best route changed in Destination object
+        if best_changed:
+            # Update prefix in the fib
+            self.fib.put_route(prefix_destination.best_route)
+            # Try to delete superfluous children
+            if not self._delete_superfluous_children(prefix_destination, children_prefixes):
+                # If children have not been deleted, update them
+                self._update_prefix_children(children_prefixes)
+
+    def _update_prefix_children(self, children_prefixes):
```

Comment:
Is this intended to trigger recalculation of the effective computed next-hops?

Reply:
Yes, calling `child_prefix_dest.best_route` the best RibRoute object is passed. After that in FIB, when `next_hops` is accessed, the recalculation is triggered each time.